pkg/ivd/ivd_protected_entity_type_manager_util.go

```diff
@@ -8,41 +8,164 @@ import (
 	"github.com/vmware/govmomi/find"
+	"github.com/vmware/govmomi/property"
 	"github.com/vmware/govmomi/vim25"
+	"github.com/vmware/govmomi/vim25/mo"
 	vim25types "github.com/vmware/govmomi/vim25/types"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/client-go/kubernetes"
 	"k8s.io/client-go/rest"
 	"strings"
 )
 
-func findAllDatastores(ctx context.Context, client *vim25.Client) ([]vim25types.ManagedObjectReference, error) {
+func findHostsOfNodeVMs(ctx context.Context, client *vim25.Client, config *rest.Config, logger logrus.FieldLogger) ([]vim25types.ManagedObjectReference, error) {
+	// #1: get hostNames of all node VMs
+	clientSet, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return nil, err
+	}
+
+	nodeList, err := clientSet.CoreV1().Nodes().List(metav1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	vmHostNameMap := make(map[string]bool)
+	for _, node := range nodeList.Items {
+		if node.Name == "" {
+			return nil, errors.New("One of the node VM in the cluster has empty node name")
+		}
+		vmHostNameMap[node.Name] = true
+	}
+
+	// #2: go through the VM list in this VC and get their host from vm.runtime
+	finder := find.NewFinder(client)
+
+	vms, err := finder.VirtualMachineList(ctx, "*")
+	if err != nil {
+		return nil, err
+	}
+
+	var vmRefList []vim25types.ManagedObjectReference
+	for _, vm := range vms {
+		vmRefList = append(vmRefList, vm.Reference())
+	}
+
+	pc := property.DefaultCollector(client)
+	var vmMoList []mo.VirtualMachine
+	err = pc.Retrieve(ctx, vmRefList, []string{"runtime", "guest"}, &vmMoList)
+	if err != nil {
+		return nil, err
+	}
+
+	var hostList []vim25types.ManagedObjectReference
+	hostRefMap := make(map[vim25types.ManagedObjectReference]bool)
+	for _, vmMo := range vmMoList {
+		_, ok := vmHostNameMap[vmMo.Guest.HostName]
+		if !ok {
+			continue
+		}
+		_, ok = hostRefMap[*vmMo.Runtime.Host]
+		if !ok {
+			hostRefMap[*vmMo.Runtime.Host] = true
+			hostList = append(hostList, *vmMo.Runtime.Host)
+		}
+	}
+
+	logger.Debugf("hostList = %v", hostList)
+	return hostList, nil
+}
+
+func findSharedDatastoresFromAllNodeVMs(ctx context.Context, client *vim25.Client, config *rest.Config, logger logrus.FieldLogger) ([]vim25types.ManagedObjectReference, error) {
 	finder := find.NewFinder(client)
+
+	hosts, err := findHostsOfNodeVMs(ctx, client, config, logger)
+	if err != nil {
+		return nil, err
+	}
+	nHosts := len(hosts)
+
 	dss, err := finder.DatastoreList(ctx, "*")
 	if err != nil {
 		return nil, err
 	}
 
 	var dsList []vim25types.ManagedObjectReference
 	for _, ds := range dss {
-		dsList = append(dsList, ds.Reference())
-	}
-	if len(dsList) == 0 {
-		return nil, errors.New("No datastore can be found")
+		dsType, err := ds.Type(ctx)
+		if err != nil {
+			logger.WithError(err).Warnf("Failed to get type of datastore %v", ds.Reference())
+			continue
+		}
+
+		if dsType == vim25types.HostFileSystemVolumeFileSystemTypeNFS41 {
+			continue
```

Comment:
Let's put a comment here as to WHY we're skipping NFS41.  Eventually we will support FCDs on 4.1

Reply:
Sure, will do.